# Max Verstappen Probability to get to current position(after Qatar GP) using Monte Carlo Method

## Lets calculate the probability that Max Verstappen still had to close the gap by atleast 92 points

In [1]:
import pandas as pd
import numpy as np
import fastf1
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [2]:
schedule_2025 = fastf1.get_event_schedule(2025, include_testing=0)
schedule_2025

req         WARNING 	DEFAULT CACHE ENABLED! (62.12 MB) C:\Users\farha\AppData\Local\Temp\fastf1


RoundNumber               Country           Location  \
1             1             Australia          Melbourne   
2             2                 China           Shanghai   
3             3                 Japan             Suzuka   
4             4               Bahrain             Sakhir   
5             5          Saudi Arabia             Jeddah   
6             6         United States      Miami Gardens   
7             7                 Italy              Imola   
8             8                Monaco             Monaco   
9             9                 Spain          Barcelona   
10           10                Canada           Montréal   
11           11               Austria          Spielberg   
12           12        United Kingdom        Silverstone   
13           13               Belgium  Spa-Francorchamps   
14           14               Hungary           Budapest   
15           15           Netherlands          Zandvoort   
16           16                 Italy              Monza   
17           17            Azerbaijan               Baku   
18           18             Singapore         Marina Bay   
19           19         United States             Austin   
20           20                Mexico        Mexico City   
21           21                Brazil          São Paulo   
22           22         United States          Las Vegas   
23           23                 Qatar             Lusail   
24           24  United Arab Emirates         Yas Island   

                                    OfficialEventName  EventDate  \
1   FORMULA 1 LOUIS VUITTON AUSTRALIAN GRAND PRIX ... 2025-03-16   
2          FORMULA 1 HEINEKEN CHINESE GRAND PRIX 2025 2025-03-23   
3          FORMULA 1 LENOVO JAPANESE GRAND PRIX 2025  2025-04-06   
4          FORMULA 1 GULF AIR BAHRAIN GRAND PRIX 2025 2025-04-13   
5         FORMULA 1 STC SAUDI ARABIAN GRAND PRIX 2025 2025-04-20   
6          FORMULA 1 CRYPTO.COM MIAMI GRAND PRIX 2025 2025-05-04   
7   FORMULA 1 AWS GRAN PREMIO DEL MADE IN ITALY E ... 2025-05-18   
8       FORMULA 1 TAG HEUER GRAND PRIX DE MONACO 2025 2025-05-25   
9         FORMULA 1 ARAMCO GRAN PREMIO DE ESPAÑA 2025 2025-06-01   
10        FORMULA 1 PIRELLI GRAND PRIX DU CANADA 2025 2025-06-15   
11     FORMULA 1 MSC CRUISES AUSTRIAN GRAND PRIX 2025 2025-06-29   
12    FORMULA 1 QATAR AIRWAYS BRITISH GRAND PRIX 2025 2025-07-06   
13   FORMULA 1 MOËT & CHANDON BELGIAN GRAND PRIX 2025 2025-07-27   
14         FORMULA 1 LENOVO HUNGARIAN GRAND PRIX 2025 2025-08-03   
15           FORMULA 1 HEINEKEN DUTCH GRAND PRIX 2025 2025-08-31   
16        FORMULA 1 PIRELLI GRAN PREMIO D’ITALIA 2025 2025-09-07   
17  FORMULA 1 QATAR AIRWAYS AZERBAIJAN GRAND PRIX ... 2025-09-21   
18  FORMULA 1 SINGAPORE AIRLINES SINGAPORE GRAND P... 2025-10-05   
19  FORMULA 1 MSC CRUISES UNITED STATES GRAND PRIX... 2025-10-19   
20  FORMULA 1 GRAN PREMIO DE LA CIUDAD DE MÉXICO 2025 2025-10-26   
21  FORMULA 1 MSC CRUISES GRANDE PRÊMIO DE SÃO PAU... 2025-11-09   
22       FORMULA 1 HEINEKEN LAS VEGAS GRAND PRIX 2025 2025-11-22   
23      FORMULA 1 QATAR AIRWAYS QATAR GRAND PRIX 2025 2025-11-30   
24  FORMULA 1 ETIHAD AIRWAYS ABU DHABI GRAND PRIX ... 2025-12-07   

                    EventName        EventFormat    Session1  \
1       Australian Grand Prix       conventional  Practice 1   
2          Chinese Grand Prix  sprint_qualifying  Practice 1   
3         Japanese Grand Prix       conventional  Practice 1   
4          Bahrain Grand Prix       conventional  Practice 1   
5    Saudi Arabian Grand Prix       conventional  Practice 1   
6            Miami Grand Prix  sprint_qualifying  Practice 1   
7   Emilia Romagna Grand Prix       conventional  Practice 1   
8           Monaco Grand Prix       conventional  Practice 1   
9          Spanish Grand Prix       conventional  Practice 1   
10        Canadian Grand Prix       conventional  Practice 1   
11        Austrian Grand Prix       conventional  Practice 1   
12         British Grand Prix     

In [3]:
race_data = []
# Races between Zandvoort (R15) and Qatar (R23) are Rounds 16 through 23
rounds_to_analyze = list(range(16, 24)) 

for r in rounds_to_analyze:
    try:
        # Load the race session
        session = fastf1.get_session(2025, r, 'R')
        session.load(telemetry=False, weather=False, messages=False)
        
        results = session.results
        
        # Get points for the drivers
        ver_points = results[results['Abbreviation'] == 'VER']['Points'].iloc[0]
        nor_points = results[results['Abbreviation'] == 'NOR']['Points'].iloc[0]
        pia_points = results[results['Abbreviation'] == 'PIA']['Points'].iloc[0]

        
        race_data.append({
            'Race': session.event['EventName'],
            'VER_Race_Points': ver_points,
            'NOR_Race_Points': nor_points,
            'PIA_Race_Points': pia_points 
        })
        
    except Exception as e:
        print(f"Skipping Round {r} due to error: {e}")

# Convert the list of dictionaries to a DataFrame
df_race = pd.DataFrame(race_data)

core           INFO 	Loading data for Italian Grand Prix - Race [v3.7.0]
req            INFO 	Using cached data for session_info
req            INFO 	Using cached data for driver_info
req            INFO 	Using cached data for session_status_data
req            INFO 	Using cached data for lap_count
req            INFO 	Using cached data for track_status_data
req            INFO 	Using cached data for _extended_timing_data
req            INFO 	Using cached data for timing_app_data
core           INFO 	Processing timing data...
core           INFO 	Finished loading data for 20 drivers: ['1', '4', '81', '16', '63', '44', '23', '5', '12', '6', '55', '87', '22', '30', '31', '10', '43', '18', '14', '27']
core           INFO 	Loading data for Azerbaijan Grand Prix - Race [v3.7.0]
req            INFO 	Using cached data for session_info
req            INFO 	Using cached data for driver_info
req            INFO 	Using cached data for session_status_data
req            INFO 	Using cached data for

In [4]:
df_race

Race  VER_Race_Points  NOR_Race_Points  PIA_Race_Points
0        Italian Grand Prix             25.0             18.0             15.0
1     Azerbaijan Grand Prix             25.0              6.0              0.0
2      Singapore Grand Prix             18.0             15.0             12.0
3  United States Grand Prix             25.0             18.0             10.0
4    Mexico City Grand Prix             15.0             25.0             10.0
5      São Paulo Grand Prix             15.0             25.0             10.0
6      Las Vegas Grand Prix             25.0              0.0              0.0
7          Qatar Grand Prix             25.0             12.0             18.0

In [5]:
sprint_data = []
# Races between Zandvoort (R15) and Qatar (R23) are Rounds 16 through 23
rounds_to_analyze = list(range(16, 24)) 

SPRINT_ROUNDS = {19, 21, 23}

for r in rounds_to_analyze:
    if r in SPRINT_ROUNDS:
        try:
            # Load the race session
            session = fastf1.get_session(2025, r, 'S')
            session.load(telemetry=False, weather=False, messages=False)
            
            results = session.results
            
            # Get points for the drivers
            ver_points = results[results['Abbreviation'] == 'VER']['Points'].iloc[0]
            nor_points = results[results['Abbreviation'] == 'NOR']['Points'].iloc[0]
            pia_points = results[results['Abbreviation'] == 'PIA']['Points'].iloc[0]

            
            sprint_data.append({
                'Race': session.event['EventName'],
                'VER_Points_S': ver_points,
                'NOR_Points_S': nor_points,
                'PIA_Points_S': pia_points 
            })
        
        except Exception as e:
            print(f"Skipping Round {r} due to error: {e}")
    else:
        ver_points, nor_points, pia_points = 0.0, 0.0, 0.0
        session = fastf1.get_session(2025, r, 'R')
        session.load(telemetry=False, weather=False, messages=False)
        sprint_data.append({
                'Race': session.event['EventName'],
                'VER_Points_S': ver_points,
                'NOR_Points_S': nor_points,
                'PIA_Points_S': pia_points 
            })
    

# Convert the list of dictionaries to a DataFrame
df_sprint = pd.DataFrame(sprint_data)

core           INFO 	Loading data for Italian Grand Prix - Race [v3.7.0]
req            INFO 	Using cached data for session_info
req            INFO 	Using cached data for driver_info
req            INFO 	Using cached data for session_status_data
req            INFO 	Using cached data for lap_count
req            INFO 	Using cached data for track_status_data
req            INFO 	Using cached data for _extended_timing_data
req            INFO 	Using cached data for timing_app_data
core           INFO 	Processing timing data...
core           INFO 	Finished loading data for 20 drivers: ['1', '4', '81', '16', '63', '44', '23', '5', '12', '6', '55', '87', '22', '30', '31', '10', '43', '18', '14', '27']
core           INFO 	Loading data for Azerbaijan Grand Prix - Race [v3.7.0]
req            INFO 	Using cached data for session_info
req            INFO 	Using cached data for driver_info
req            INFO 	Using cached data for session_status_data
req            INFO 	Using cached data for

In [6]:
df_sprint

Race  VER_Points_S  NOR_Points_S  PIA_Points_S
0        Italian Grand Prix           0.0           0.0           0.0
1     Azerbaijan Grand Prix           0.0           0.0           0.0
2      Singapore Grand Prix           0.0           0.0           0.0
3  United States Grand Prix           8.0           0.0           0.0
4    Mexico City Grand Prix           0.0           0.0           0.0
5      São Paulo Grand Prix           5.0           8.0           0.0
6      Las Vegas Grand Prix           0.0           0.0           0.0
7          Qatar Grand Prix           5.0           6.0           8.0

In [7]:
df_sprint_points_only = df_sprint[['Race', 'VER_Points_S', 'NOR_Points_S', 'PIA_Points_S']]

df_weekend = df_race.merge(
    df_sprint_points_only, 
    on='Race', 
    how='left'
)

df_weekend[['VER_Points_S', 'NOR_Points_S', 'PIA_Points_S']] = \
    df_weekend[['VER_Points_S', 'NOR_Points_S', 'PIA_Points_S']].fillna(0.0)

df_weekend['VER_Total_Points'] = df_weekend['VER_Race_Points'] + df_weekend['VER_Points_S']
df_weekend['NOR_Total_Points'] = df_weekend['NOR_Race_Points'] + df_weekend['NOR_Points_S']
df_weekend['PIA_Total_Points'] = df_weekend['PIA_Race_Points'] + df_weekend['PIA_Points_S']

In [8]:
df_weekend['Points_Difference_VER_NOR'] = df_weekend['VER_Total_Points'] - df_weekend['NOR_Total_Points']

In [9]:
df_weekend

Race  VER_Race_Points  NOR_Race_Points  \
0        Italian Grand Prix             25.0             18.0   
1     Azerbaijan Grand Prix             25.0              6.0   
2      Singapore Grand Prix             18.0             15.0   
3  United States Grand Prix             25.0             18.0   
4    Mexico City Grand Prix             15.0             25.0   
5      São Paulo Grand Prix             15.0             25.0   
6      Las Vegas Grand Prix             25.0              0.0   
7          Qatar Grand Prix             25.0             12.0   

   PIA_Race_Points  VER_Points_S  NOR_Points_S  PIA_Points_S  \
0             15.0           0.0           0.0           0.0   
1              0.0           0.0           0.0           0.0   
2             12.0           0.0           0.0           0.0   
3             10.0           8.0           0.0           0.0   
4             10.0           0.0           0.0           0.0   
5             10.0           5.0           8.0           0.0   
6              0.0           0.0           0.0           0.0   
7             18.0           5.0           6.0           8.0   

   VER_Total_Points  NOR_Total_Points  PIA_Total_Points  \
0              25.0              18.0              15.0   
1              25.0               6.0               0.0   
2              18.0              15.0              12.0   
3              33.0              18.0              10.0   
4              15.0              25.0              10.0   
5              20.0              33.0              10.0   
6              25.0               0.0               0.0   
7              30.0              18.0              26.0   

   Points_Difference_VER_NOR  
0                        7.0  
1                       19.0  
2                        3.0  
3                       15.0  
4                      -10.0  
5                      -13.0  
6                       25.0  
7                       12.0

Now extract the necessary data from the dataframe

In [ ]:
mean_point_difference = df_weekend['Points_Difference_VER_NOR'].mean()
std_point_difference = df_weekend['Points_Difference_VER_NOR'].std()
target_gap = 92 #104 - 12
no_races = 8 #Italy to Qatar
no_simulations = 100000 

In [12]:
points_diffs = np.random.normal(loc=mean_point_difference, scale=std_point_difference, size=(no_simulations, no_races))

In [ ]:
total_gains = points_diffs.sum(axis=1)
successful_simulations = (total_gains >= target_gap).sum()
probability = successful_simulations / no_simulations

In [16]:
probability

np.float64(0.18416)

In [17]:
print(f"--- Monte Carlo Simulation Results ---")
print(f"Input Distribution: Normal (mu={mean_point_difference}, sigma={std_point_difference})")
print(f"Target Gain Required: {target_gap} points over {no_races} races.")
print("-" * 40)
print(f"Total Simulations Run: {no_simulations}")
print(f"Successful Scenarios (Gain >= {target_gap} pts): {successful_simulations}")
print(f"Calculated Probability: {probability:.4f}")
print(f"The percentage chance of these events or better events occur is: {probability * 100}%")

--- Monte Carlo Simulation Results ---
Input Distribution: Normal (mu=7.25, sigma=13.424391446701996)
Target Gain Required: 92 points over 8 races.
----------------------------------------
Total Simulations Run: 100000
Successful Scenarios (Gain >= 92 pts): 18416
Calculated Probability: 0.1842
The percentage chance of these events or better events occur is: 18.416%


Therefore, the chance that Max Verstappen would have reduced the gap by 92 points was calculated to be approximately 18% To put that into perspective, the odds of this happening were approximately 1 in 5.4. This means that in nearly 82 out of every 100 instances modeled under the same performance distribution (mean = 7.25 points and standard deviation approx. = 13.42 points per race), the required gap reduction would not have been achieved. Max Verstappen has overcome massive statistical odds to have one last shot at securing his 5th consecutive WDC